### PR TITLE
Add selinux resource support for modules and booleans

### LINKED ISSUE #534: Resources to support SElinux
We should add an selinux resource that can check the status of selinux configs, mode, status and perhaps ones for selinux_rule as well?

### PR DESCRIPTION
**Pull request by Vasu1105**
## Description

The test which are run manually against Centos 8 and RHEL 8

```
describe selinux do
     it { should be_disabled }
     it { should_not be_disabled }
     it { should be_enforcing }
     it { should_not be_permissive }
     it { should be_installed }
  end

  describe selinux.modules.where(status: "installed") do
    it { should exist }
    its('count') { should cmp 404 }
  end

  describe selinux.modules.where(name: "zebra") do
    it { should be_installed }
    it { should_not be_enabled }
  end

  describe selinux.modules.where(name: "xen") do
    it { should be_installed }
    it { should be_enabled }
  end

  describe selinux.modules.where(name: "test") do
    it { should be_installed }
    it { should be_enabled }
  end

  describe selinux.booleans.where(name: "httpd_enable_homedirs") do
    it { should_not be_on }
    its('states') { should_not cmp "on" }
    its('defaults') { should_not cmp "on" }
  end

  describe selinux.booleans.where(name: "xend_run_blktap") do
    it { should be_on }
    its('states') { should cmp "on" }
    its('defaults') { should cmp "on" }
  end

  describe selinux.booleans.where(name: "xend_run_blktap", state: "on") do
    it { should exist }
  end

  describe selinux.booleans do
    its("names") { should include "httpd_enable_homedirs" }
  end

  describe selinux do
    its('policy') { should eq "targeted"}
  end

  describe selinux.booleans.where { name == "xend_run_blktap" && state == "on" } do
   it { should exist }
  end

  describe selinux.modules.where("zebra") do
    it { should exist }
    it { should be_installed }
    it { should be_enabled }
  end
```

## Related Issue

Fix #534 
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New content (non-breaking change)
- [ ] Breaking change (a content change which would break existing functionality or processes)

## Checklist:

- [ ] I have read the **CONTRIBUTING** document.

### CODE REVIEW

File: lib/inspec/resources/selinux.rb [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+    attr_reader :modules
+    def initialize(modules)
+      @modules = modules
+    end
+
+    def to_s
+      "SElinux modules"
+    end
+  end
```
Comment (clintoncwolfe) [inline lib/inspec/resources/selinux.rb]: This looks great - just need to ad docs and tests!

Comment (Vasu1105) [reply lib/inspec/resources/selinux.rb]: Docs and test is added. Thanks @clintoncwolfe .

**Comment by Vasu1105** [comment]
@clintoncwolfe  could you please have re-look at this one as I have added docs and test on this.

File: docs-chef-io/content/inspec/resources/selinux.md [hunk 2]
```diff
@@ -11,9 +11,9 @@ platform = "linux"
     parent = "inspec/resources/os"
 +++
 
-Use the `selinux` Chef InSpec audit resource to test the state and mode of SELinux policy.
+Use the `selinux` Chef Inspec audit resource to test the configuration data of the SELinux policy, SELinux modules and SELinux booleans.
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/selinux.md]:
```suggestion
Use the `selinux` Chef InSpec audit resource to test the configuration data of the SELinux policy, SELinux modules, and SELinux booleans.
```

File: docs-chef-io/content/inspec/resources/selinux.md [hunk 3]
```diff
@@ -11,9 +11,9 @@ platform = "linux"
...
 
-The `selinux` resource extracts and exposes data reported by the `sestatus` command.
+The `selinux` resource extracts and exposes data reported by the `sestatus`, `semodule -lfull`and `semanage boolean -l -n` command.
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/selinux.md]:
```suggestion
The `selinux` resource extracts and exposes data reported by the `sestatus`, `semodule -lfull`, and `semanage boolean -l -n` command.
```

File: docs-chef-io/content/inspec/resources/selinux.md [hunk 4]
```diff
@@ -34,6 +36,41 @@ The `selinux` Chef InSpec resource block tests the state and mode of SELinux pol
       it { should_not be_permissive }
     end
 
+The `selinux` resource block also declares allows you to write test for many modules:
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/selinux.md]:
```suggestion
The `selinux` resource block allows you to write tests for multiple modules:
```

File: docs-chef-io/content/inspec/resources/selinux.md [hunk 5]
```diff
@@ -34,6 +36,41 @@ The `selinux` Chef InSpec resource block tests the state and mode of SELinux pol
...
+
+    describe selinux.modules.where("zebra") do
+      it { should exist }
+      it { should be_installed }
+      it { should be_enabled }
+    end
+
+or:
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/selinux.md:47]:
```suggestion
or:

```

File: docs-chef-io/content/inspec/resources/selinux.md [hunk 6]
```diff
@@ -34,6 +36,41 @@ The `selinux` Chef InSpec resource block tests the state and mode of SELinux pol
...
+    describe selinux.modules.where(status: "installed") do
+      it { should exist }
+      its('count') { should cmp 404 }
+    end
+
+where
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/selinux.md]:
```suggestion
where:
```

File: docs-chef-io/content/inspec/resources/selinux.md [hunk 7]
```diff
@@ -34,6 +36,41 @@ The `selinux` Chef InSpec resource block tests the state and mode of SELinux pol
...
+
+- `.where()` may specify a specific item and value, to which the resource parameters are compared
+- `name`, `status`, `state`, `priority` are valid parameters for `modules`
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/selinux.md]:
```suggestion
- `.where()` specifies the parameter and expected value.
- `name`, `status`, `state`, and `priority` are valid parameters.
```

File: docs-chef-io/content/inspec/resources/selinux.md [hunk 8]
```diff
@@ -34,6 +36,41 @@ The `selinux` Chef InSpec resource block tests the state and mode of SELinux pol
...
+
+The `selinux` resource block also declares allows you to write test for many booleans:
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/selinux.md]:
```suggestion
The `selinux` resource block allows you to write tests for multiple booleans:
```

File: docs-chef-io/content/inspec/resources/selinux.md [hunk 9]
```diff
@@ -34,6 +36,41 @@ The `selinux` Chef InSpec resource block tests the state and mode of SELinux pol
...
+
+or:
+
+  describe selinux.booleans.where(name: "xend_run_blktap", state: "on") do
+    it { should exist }
+    its('defaults') { should cmp "on" }
+  end
+
+- `.where()` may specify a specific item and value, to which the resource parameters are compared
+- `name`, `state`, `default` are valid parameters for `booleans`
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/selinux.md]:
```suggestion
- `.where()`  specifies the parameter and expected value.
- `name`, `state`, and `default` are valid parameters for `booleans`.
```

File: docs-chef-io/content/inspec/resources/selinux.md [hunk 10]
```diff
@@ -52,13 +89,18 @@ describe selinux do
   it { should be_enforcing }
 end
 
+### Test if selinux policy type
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/selinux.md]:
```suggestion
### Test the SELinux policy type

```

File: docs-chef-io/content/inspec/resources/selinux.md [hunk 11]
```diff
@@ -52,13 +89,18 @@ describe selinux do
...
+end
+
 ## Matchers
 
 For a full list of available matchers, please visit our [matchers page](/inspec/matchers/).
 
 ### be_installed
 
-The `be_installed` matcher tests if the SELinux is installed on the system:
+The `be_installed` matcher tests if the SElinux policy or SElinux modules is installed on the system:
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/selinux.md]:
```suggestion
The `be_installed` matcher tests if the SElinux policy or SElinux modules are installed on the system:
```

File: docs-chef-io/content/inspec/resources/selinux.md [hunk 12]
```diff
@@ -79,3 +121,48 @@ The `be_enforcing` matcher tests if the SELinux mode is set to enforcing:
...
+
+The `be_enabled` matcher tests if the SElinux module is enabled:
+
+    it { should be_enabled }
+
+## Resource Parameters
+
+- `names`, `status`, `states`, `priorities`,  are valid parameters for SELinux `modules`
+
+- `names`, `status`, `states`, `defaults`,  are valid parameters for SELinux `booleans`
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/selinux.md]:
```suggestion
- `names`, `status`, `states`, and `priorities` are valid parameters for SELinux policy modules.

- `names`, `status`, `states`, and `defaults` are valid parameters for SELinux booleans.
```

File: docs-chef-io/content/inspec/resources/selinux.md [hunk 13]
```diff
@@ -79,3 +121,48 @@ The `be_enforcing` matcher tests if the SELinux mode is set to enforcing:
...
+
+## Resource Parameter Examples
+
+### modules
+
+`modules` returns the information about SELinux modules as returned by [semodule -lfull](https://man7.org/linux/man-pages/man8/semodule.8.html).
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/selinux.md]:
```suggestion
`modules` returns information about SELinux modules using the [semodule -lfull](https://man7.org/linux/man-pages/man8/semodule.8.html) command.
```

File: docs-chef-io/content/inspec/resources/selinux.md [hunk 14]
```diff
@@ -79,3 +121,48 @@ The `be_enforcing` matcher tests if the SELinux mode is set to enforcing:
...
+
+Note: The `semodule -l` command does not provide `version` information in newer versions of Linux based systems like RHEL8 and Centos8 so we are not supporting that option [REF](https://access.redhat.com/solutions/2760071).
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/selinux.md]:
```suggestion
Note: The `semodule -l` command [does not provide version information](https://access.redhat.com/solutions/2760071) for newer versions of Linux-based systems like RHEL8 and Centos8, so we do not support that option.
```

File: docs-chef-io/content/inspec/resources/selinux.md [hunk 15]
```diff
@@ -79,3 +121,48 @@ The `be_enforcing` matcher tests if the SELinux mode is set to enforcing:
...
+
+describe selinux.modules do
+  its("names") { should include "zebra" }
+  its("status") { should include "installed" }
+  its("states") { should include "enabled" }
+  its("priorities") { should include "100" }
+end
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/selinux.md:159]:
````suggestion
```ruby
describe selinux.modules do
  its("names") { should include "zebra" }
  its("status") { should include "installed" }
  its("states") { should include "enabled" }
  its("priorities") { should include "100" }
end
```
````

File: docs-chef-io/content/inspec/resources/selinux.md [hunk 16]
```diff
@@ -79,3 +121,48 @@ The `be_enforcing` matcher tests if the SELinux mode is set to enforcing:
...
+
+### booleans
+
+`booleans` returns the information about SELinux booleans as returned by [semanage boolean -l -n](https://man7.org/linux/man-pages/man8/semanage-boolean.8.html)
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/selinux.md]:
```suggestion
`booleans` returns information about SELinux booleans using the [semanage boolean -l -n](https://man7.org/linux/man-pages/man8/semanage-boolean.8.html) command.
```

File: docs-chef-io/content/inspec/resources/selinux.md [hunk 17]
```diff
@@ -79,3 +121,48 @@ The `be_enforcing` matcher tests if the SELinux mode is set to enforcing:
...
+
+describe selinux.booleans do
+  its("names") { should include "httpd_enable_homedirs" }
+  its("states") { should include "on" }
+  its("states") { should include "off" }
+  its("defaults") { should include "on" }
+  its("defaults") { should include "off" }
+end
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/resources/selinux.md:172]:
````suggestion
```ruby
describe selinux.booleans do
  its("names") { should include "httpd_enable_homedirs" }
  its("states") { should include "on" }
  its("states") { should include "off" }
  its("defaults") { should include "on" }
  its("defaults") { should include "off" }
end
```
````

File: lib/inspec/resources/selinux.rb [hunk 18]
```diff
@@ -1,11 +1,54 @@
...
+      "SElinux booleans"
+    end
+  end
+
   class Selinux < Inspec.resource(1)
     name "selinux"
     supports platform: "linux"
 
-    desc "Use selinux Inspec resource to test state/mode of the selinux policy."
+    desc "Use selinux Chef Inspec resource to test the configuration data of the selinux policy, selinux modules and selinux booleans."
```
Comment (IanMadd) [inline lib/inspec/resources/selinux.rb]:
```suggestion
    desc "Use the selinux Chef InSpec resource to test the configuration data of the SELinux policy, SELinux modules, and SELinux booleans."
```

**Review by IanMadd — COMMENTED** [review]
A bunch of docs suggestions.

**Comment by Vasu1105** [comment]
@IanMadd I have updated the docs as per review comments.

**Pull request merged**